Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/PrefabPublicHandler.cpp

```diff
@@ -1311,10 +1283,116 @@ namespace AzToolsFramework
+                RemoveLink(outInstance, commonOwningTemplateId, undoBatch.GetUndoBatch());
+                outInstance.reset();
+            }
+
+            // Step 2 - Entity Removals.
+            AZStd::vector<PrefabDom> removedEntityDomList;
+            AZStd::vector<AZStd::string> removedEntityAliasPathList;
+            AZStd::unordered_set<AZ::EntityId> entitiesThatGotRemoved;
+
+            for (const AZ::Entity* entity : entityList)
+            {
+                const AZ::EntityId entityId = entity->GetId();
+                const AZStd::string entityAliasPath = m_instanceToTemplateInterface->GenerateEntityAliasPath(entityId);
+
+                // If parent is valid and not seen before (because we only need to capture the initial state),
+                // then get the before state of the parent entity DOM.
+                AZ::EntityId parentEntityId;
+                AZ::TransformBus::EventResult(parentEntityId, entityId, &AZ::TransformBus::Events::GetParentId);
+                if (parentEntityId.IsValid() &&
+                    parentEntityDomBeforeRemovalMap.find(parentEntityId) == parentEntityDomBeforeRemovalMap.end())
+                {
+                    AZStd::string parentEntityAlias = m_instanceToTemplateInterface->GenerateEntityAliasPath(parentEntityId);
+                    PrefabDomPath parentEntityAliasInOwningTemplatePath(parentEntityAlias.c_str());
+                    PrefabDomValue* parentEntityDomPtr = parentEntityAliasInOwningTemplatePath.Get(commonOwningTemplateDom);
+                    if (parentEntityDomPtr)
+                    {
+                        // Deep copy is needed to capture the current DOM state.
+                        parentEntityDomBeforeRemovalMap[parentEntityId] = PrefabDom();
+                        PrefabDom& domInMap = parentEntityDomBeforeRemovalMap[parentEntityId];
+                        domInMap.CopyFrom(*parentEntityDomPtr, domInMap.GetAllocator());
+                    }
+                }
+
+                // Detaches the entity.
+                commonOwningInstance->get().DetachEntity(entityId).release();
+                AZ::ComponentApplicationBus::Broadcast(&AZ::ComponentApplicationRequests::DeleteEntity, entityId);
+
+                // Captures the entity DOM from the owning template DOM for undo purpose and creates an undo node
+                // to update the template.
+                PrefabDomPath entityDomInOwningTemplatePath(entityAliasPath.c_str());
+                PrefabDomValue* removedEntityDomPtr = entityDomInOwningTemplatePath.Get(commonOwningTemplateDom);
+                removedEntityDomList.emplace_back();
+                removedEntityAliasPathList.push_back(AZStd::move(entityAliasPath));
+                if (removedEntityDomPtr)
+                {
+                    PrefabDom& domAdded = removedEntityDomList.back();
+                    domAdded.CopyFrom(*removedEntityDomPtr, domAdded.GetAllocator());
```

Comment:
same here. Declare the dom first, copy into it and insert into the map at the end.

Reply:
This will be changed to pointer.

The reason I was doing here is the same: avoid copying or (even moving) PrefabDom.

```cpp
AZStd::vector<PrefabDom> domList;
PrefabDom dom;
domList.push_back(AZStd::move(dom));   // rapidjson's Move() does not work because it returns PrefabDomVaule.
```
Move constructor/assignment of rapidjson::Document are defined here: https://rapidjson.org/document_8h_source.html (Line 2518)